woocommerce/woocommerce-ios
# [Orders with Coupons] Fix icon tooltip rendering when order has no discounts

https://github.com/woocommerce/woocommerce-ios/pull/10956 should be merged first

## Description
This PR addresses a bug when rendering the tooltip icon on product discounts, which should appear only when an Order has coupons applied to it (do product discounts are disallowed and we show the informational tooltip to offer more info about it):

<img width="416" alt="Screenshot 2023-10-17 at 18 03 24" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/d67da323-4c34-4da5-a740-9a82cba0a87f">


## Changes
* The fix is only the line here: https://github.com/woocommerce/woocommerce-ios/commit/cafdf252fe0658824dda5f3321960735e1c953b6 . We check if discounts are disallowed, and if that's the case then we render the tooltip icon.
* Additionally we do some clean-up by:
  * We declare a new property `shouldDisallowDiscounts` that acts as a wrapper for `orderHasCoupons'. This encapsulates and hides the implementation details, and clarifies the logic. 
  * Add unit tests to both `EditableOrderViewModelTests` and `ProductRowViewModelTests`.
  * Extracting `TooltipView` to its own file.

## Testing instructions
* Go to Orders > + > `+ Add Product` > Select a product > Tap on the chevron icon to expand the product details.
* Observe that the `+ Add discount` row appears without informational tooltip
* Tap on `+ Add Coupon` and apply a valid coupon to the order
* Observe that the `+ Add discount` row is disabled and an informational tooltip can be seen and tapped.

| No discounts or coupons | Coupon applied |
|--------|--------|
| ![_Case 1 - No discounts or coupons - Icon hidden](https://github.com/woocommerce/woocommerce-ios/assets/3812076/abd64839-60e1-476f-a2c1-a78bf5b10b66) | ![_Case 3 - No discounts but a coupons - Icon appear](https://github.com/woocommerce/woocommerce-ios/assets/3812076/18a3c0ff-3037-4120-8707-6b6bf41747f1) |
